src/helpers/users.js

```diff
@@ -278,9 +279,37 @@ exports.updateUserInactivityDate = async (user, contractEndDate, credentials) =>
 
 exports.removeUser = async (user, credentials) => {
   if (UtilsHelper.areObjectIdsEquals(user._id, credentials._id)) {
+    const hasActivityHistories = await ActivityHistory.countDocuments({ user: user._id }, { limit: 1 });
     await CompanyLinkRequest.deleteOne({ user: user._id });
-    await Course.updateMany({ trainees: user._id }, { $pull: { trainees: user._id } });
-    await User.deleteOne({ _id: user._id });
+    if (hasActivityHistories) {
+      const fieldsToUnset = { 'local.password': '', 'identity.firstname': '', passwordToken: '' };
+      const fieldsToKeep = ['local', 'identity', 'origin', 'serialNumber', '_id', 'company', 'createdAt', 'updatedAt'];
+
+      for (const key of Object.keys(user)) {
+        if (!fieldsToKeep.includes(key)) {
+          const entry = { [key]: '' };
+          Object.assign(fieldsToUnset, entry);
+        }
+      }
+      await User.updateOne(
+        { _id: user._id },
+        {
+          $unset: fieldsToUnset,
```

Comment:
Pourquoi ne pas avoir mis directement: `$unset: { 'local.password': '', 'identity.firstname': '', passwordToken: '' }` ?

Reply:
parce qu'il y a d'autres champs. Quelques lignes plus bas on fait `Object.assign(fieldsToUnset, entry);` qui rajoute d'autres champs à unset dans fieldsToUnset

Reply:
En fait je ne comprends pas pourquoi on a besoin de faire cette boucle for qui ajoute les champs a `fieldsToUnset` et qu'on ne les définis pas directement dans l'objet `fieldsToUnset` ?

Reply:
parce que j'ai pas en tête tous les champs de l'objet user qu'on veut supprimer et pour être sûr de tous les avoir. + le fait que ça me paraissait plus propre qu'une liste de tous les champs.